webapp/lib/firebase_tools.dart

```diff
@@ -64,27 +70,39 @@ Future<List<Scheme>> loadSchemes(String datasetId) async {
   return ret;
 }
 
-Future<List<Message>> loadMessages(String datasetId) async {
-  List<Message> ret = <Message>[];
-
-  log.verbose("loadMessages: Loading messages for: $datasetId");
-
-  var messagesCollectionRoot = "/datasets/$datasetId/messages";
-  log.verbose("loadMessages: Root of query: $messagesCollectionRoot");
-
-  var messagesQuery = await _firestoreInstance.collection(messagesCollectionRoot).get();
-  log.verbose("loadMessages: Query constructed");
-
-  messagesQuery.forEach((message) {
-    log.verbose('loadMessages: Processing ${message.id}');
-    ret.add(new Message.fromFirebaseMap(message.data()));
+void setupListenerForFirebaseMessageUpdates(Dataset dataset, MessageUpdatesListener listener) {
+  log.verbose("setupListenerForFirebaseMessageUpdates: Loading messages for: ${dataset.id}");
+
+  var messagesCollectionRoot = "/datasets/${dataset.id}/messages";
+  log.verbose("setupListenerForFirebaseMessageUpdates: Root of query: $messagesCollectionRoot");
+
+  _firestoreInstance.collection(messagesCollectionRoot).onSnapshot.listen((querySnapshot) {
+    // No need to process local writes to Firebase
+    if (querySnapshot.metadata.hasPendingWrites) {
+      log.verbose("setupListenerForFirebaseMessageUpdates: Skipping processing of local messages");
+      return;
+    }
+
+    log.verbose("setupListenerForFirebaseMessageUpdates: Starting processing ${querySnapshot.size} messages.");
+    List<Message> added = [];
+    List<Message> modified = [];
+    querySnapshot.docChanges().forEach((documentChange) {
+      Message message = new Message.fromFirebaseMap(documentChange.doc.data());
+      if (documentChange.type == "added") {
+        added.add(message);
+      } else if (documentChange.type == "modified") {
+        modified.add(message);
+      } else {} // skip anything else
```

Comment:
The only other thing that the documentation mentions is "removed". I've added instead of this comment a logging call, which should help us notice if something like this occurs. We may want to turn that simple logging call in a log warning.